services/QuillLMS/spec/models/concerns/teacher_spec.rb

```diff
@@ -547,19 +547,14 @@
       let!(:teacher) { create(:teacher, :with_classrooms_students_and_activities) }
       let!(:assigned_units) { teacher.assigned_students_per_activity_assigned }
 
-      it 'should return the correct number of students assigned' do
-
+      it 'should return the correct students assigned' do
         assigned_units_arr = assigned_units.to_a
-        teacher.classroom_units.each { |cu|
-          cu.unit.unit_activities.each { |ua|
+        teacher.classroom_units.each do |cu|
+          cu.unit.unit_activities.each do |ua|
             matching_element = assigned_units.detect {|u| u.id == ua.activity_id && u.assigned_student_ids == cu.assigned_student_ids}
-            expect(matching_element).to_not be_nil
             expect(matching_element.assigned_student_ids).to eq(cu.assigned_student_ids)
-            expect(matching_element.created_at).to eq(ua.created_at)
-            assigned_units_arr.delete_at(assigned_units_arr.index(matching_element))
-          }
-        }
-        expect(assigned_units_arr).to be_empty
```

Comment:
This expectation was testing the spec itself, not the intended underlying method

Reply:
While I agree there might be a race condition as you described, I think we still need something related to the `delete_at` and subsequent `be_empty` test since that ensures assigned_units doesn't contain any other values outside of those in the loop.

In other words,  if we had something like `assigned_units += [some new unit]` this test should still pass.  What about just adding a counter?

```
matching_elements_count = 0

teacher.classroom_units.each do |cu|
  cu.unit.unit_activities.each do |ua|
    matching_element = assigned_units.detect {|u| u.id == ua.activity_id && u.assigned_student_ids == cu.assigned_student_ids}
    expect(matching_element.assigned_student_ids).to eq(cu.assigned_student_ids)
    matching_elements_count += 1
  end
end

expect(matching_elements_count).to eq assigned_units.countend
```

Reply:
Ok, I incorporated the spirit of your feedback, with a different implementation